### PR TITLE
[v1.8] daemon: Avoid blocking datapath on node discovery

### PR DESCRIPTION
**Pull request by christarazi**
Previously, the datapath relied on node discovery completing. With a
kvstore configured, this meant that node registration would also need to
complete first.

If the kvstore is deployed as pods intended to be managed by Cilium,
then this creates a chicken and egg problem. The kvstore pod cannot come
online because the datapath is blocked. The datapath is blocked because
the agent cannot register the node into the kvstore. And finally, the
node cannot be registered into kvstore because kvstore is not online.

Here's how the issue manifested itself:

```
$ kubectl -n cilium describe pods etcd-operator-59cf4cfb7c-288qx
Name:           etcd-operator-59cf4cfb7c-288qx
Namespace:      cilium
Priority:       0
Node:           gke-chris-form3-cluster-default-pool-3f30c3b5-zd0f/10.138.0.10
Start Time:     Fri, 15 Jan 2021 14:56:24 -0800
Labels:         io.cilium/app=etcd-operator
                pod-template-hash=59cf4cfb7c
...
Events:
  Type     Reason                  Age   From               Message
  ----     ------                  ----  ----               -------
  Normal   Scheduled               83s   default-scheduler  Successfully assigned cilium/etcd-operator-59cf4
  Warning  FailedCreatePodSandBox  22s   kubelet            Failed create pod sandbox: rpc error: code = Unk
5bf622367" network for pod "etcd-operator-59cf4cfb7c-288qx": networkPlugin cni failed to set up pod "etcd-op
ent client after 30.000000 seconds timeout: Get "http:///var/run/cilium/cilium.sock/v1/config": dial unix /v
Is the agent running?
  Normal  SandboxChanged  21s  kubelet  Pod sandbox changed, it will be killed and re-created.
```

To break the chicken and egg problem with the datapath, we need to split
the "node discovery" (`(*NodeDiscovery).StartDiscovery()`) into two
separate phases: (1) local node state is initialized and (2) node
registration which includes syncing to the kvstore if that's configured.
The `(*NodeDiscovery).Registered` channel will close when (2) completes.
The new channel `(*NodeDiscovery).LocalStateInitialized` will close when
(1) completes. This is because the datapath only relies on (1) being
complete and doesn't depend on (2).

This will unblock the datapath from the implicit dependency on the
kvstore and allow the kvstore pods to come online (Cilium generates
endpoints for them), while node registration into the kvstore continues
on in the background.

Fixes: 43997f5f0 ("loader: Wait for node configuration to generate datapath")
Related: 7045103aa ("daemon: Move KVStore initialization earlier")

Co-authored-by: André Martins <andre@cilium.io>
Co-authored-by: Joe Stringer <joe@cilium.io>
Co-authored-by: Paul Chaignon <paul@cilium.io>
Co-authored-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

```release-note
Fix bug Cilium hangs with kvstore configured
```

### CODE REVIEW

**Review by joestringer — APPROVED** [review]
Thanks for the fix! Not sure exactly who to attribute in the commit message given that @aanm, @pchaigno , @kkourt and @christarazi were all helping to pinpoint the issue but the fix itself LGTM (+ @christarazi and I manually reproduced the issue in a cluster together to confirm the fix).

**Comment by christarazi** [comment]
> Thanks for the fix! Not sure exactly who to attribute in the commit message given that @aanm, @pchaigno , @kkourt and @christarazi were all helping to pinpoint the issue but the fix itself LGTM (+ @christarazi and I manually reproduced the issue in a cluster together to confirm the fix).

Easy fix, everyone. :1st_place_medal: :slightly_smiling_face:

**Comment by christarazi** [comment]
test-backport-1.8

**Comment by christarazi** [comment]
Travis hit https://github.com/cilium/cilium/issues/12756

**Comment by christarazi** [comment]
test-upstream-k8s

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
Changed release note from bug to misc since it doesn't affect s released version

**Comment by brb** [comment]
@christarazi Just to double check, but with this patch the datapath is still loaded after https://github.com/cilium/cilium/blob/master/pkg/k8s/init.go#L200?

**Comment by aanm** [comment]
@christarazi I've ported the PR into the v1.9 branch https://github.com/cilium/cilium/pull/14629

**Comment by christarazi** [comment]
> @christarazi Just to double check, but with this patch the datapath is still loaded after https://github.com/cilium/cilium/blob/master/pkg/k8s/init.go#L200?

@brb Correct, the datapath is loaded after `WaitForNodeInformation`

**Comment by joestringer** [comment]
Dropped regression since this doesn't affect any released versions.

**Comment by pchaigno** [comment]
Do we need a forward-port in master as well? (I can take care of it.)

In general I'm not sure I understand why we sometimes open PRs against branches before forward-porting. Wouldn't it be easier to open against master and backport after? It doesn't block hotfix and would better fit our usual workflow.

**Comment by christarazi** [comment]
> Do we need a forward-port in master as well? (I can take care of it.)

Yes :+1: 

> In general I'm not sure I understand why we sometimes open PRs against branches before forward-porting. Wouldn't it be easier to open against master and backport after? It doesn't block hotfix and would better fit our usual workflow.

I agree with you in general. I think sometimes, for me at least, when generating a hotfix, it can be faster to resolve the issue on the affected branch immediately, than to create it on master.  Creating a backport PR before the master PR is merged can be slightly more error-prone work IMO. We'd have to manually cherry-pick the commit, remember to update the upstream git SHA of the commit if there were code review comments to address, and remember to address the code reviews comments on all branches to backport. I recall a few instances where we forget to do one of those things. It's nothing that a followup PR can't fix, but at least these are my reasons to avoid doing it. Those are the things I consider when choosing to do this or waiting until the master PR can be merged, if I feel that a hotfix is slightly more urgent.

**Comment by aanm** [comment]
> @pchaigno
> 
> > Do we need a forward-port in master as well? (I can take care of it.)
> 
> Yes 
> 
> > In general I'm not sure I understand why we sometimes open PRs against branches before forward-porting. Wouldn't it be easier to open against master and backport after? It doesn't block hotfix and would better fit our usual workflow.
> 
> I agree with you in general. I think sometimes, for me at least, when generating a hotfix, it can be faster to resolve the issue on the affected branch immediately, than to create it on master. Creating a backport PR before the master PR is merged can be slightly more error-prone work IMO. We'd have to manually cherry-pick the commit, remember to update the upstream git SHA of the commit if there were code review comments to address, and remember to address the code reviews comments on all branches to backport. I recall a few instances where we forget to do one of those things. It's nothing that a followup PR can't fix, but at least these are my reasons to avoid doing it. Those are the things I consider when choosing to do this or waiting until the master PR can be merged, if I feel that a hotfix is slightly more urgent.

@pchaigno @christarazi That's it true but keep in mind doing that makes it difficult to keep track which commits have been backported from master to a particular branch.

**Pull request merged**
